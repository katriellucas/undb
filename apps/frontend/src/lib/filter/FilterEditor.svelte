<script lang="ts">
	import type { IFilter, IQueryFieldSchema } from '@undb/core'
	import Sortable, { type SortableEvent } from 'sortablejs'
	import { identity, isNumber } from 'lodash-es'
	import FilterItem from '$lib/table/FilterItem.svelte'

	export let value: Partial<IFilter>[] = []
	export let readonly = false
	export let fieldFilter: (field: IQueryFieldSchema) => boolean = identity
	export let createInitial = true

	$: if (!value) value = []

	$: if (!value?.length && !readonly && createInitial) {
		add()
	}

	const TEMP_ID = '__TEMP_ID'

	const add = () => {
		if (value.some((v) => v.path === TEMP_ID)) return
		value = [...value, { path: TEMP_ID }]
	}

	const remove = (index: number) => {
		value = value.filter((f, i) => i !== index)
	}

	const onEnd = (event: SortableEvent) => {
		const { oldIndex, newIndex } = event
		if (isNumber(oldIndex) && isNumber(newIndex)) {
			;[value[oldIndex], value[newIndex]] = [value[newIndex], value[oldIndex]]
		}
	}

	let el: HTMLUListElement
	$: if (el) {
		Sortable.create(el, {
			animation: 200,
			direction: 'vertical',
			onEnd,
			handle: '.handle',
		})
	}
</script>

{#if value?.length}
	<ul class="space-y-2" bind:this={el}>
		{#each value as filter, index (String(filter.path) + filter.operator + String(filter.value))}
			<FilterItem {filter} {index} {remove} {readonly} {fieldFilter} />
		{/each}
	</ul>
{:else}
	<slot name="empty" />
{/if}
{#if !readonly}
	<slot {add} />
{/if}
