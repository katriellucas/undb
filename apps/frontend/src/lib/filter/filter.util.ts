import { isOperatorWithoutValue, type IFilter } from '@undb/core'
import { isNil } from 'lodash-es'

export const getValidFilters = (value: Partial<IFilter>[] = []) =>
	value.filter(
		(v) => !!v.path && !!v.operator && !!v.type && (isOperatorWithoutValue(v.operator) ? true : !isNil(v.value)),
	) as IFilter[]
