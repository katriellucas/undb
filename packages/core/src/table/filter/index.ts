export * from './conjunction.js'
export * from './filter.js'
export * from './operators.js'
export * from './root-filter.js'
