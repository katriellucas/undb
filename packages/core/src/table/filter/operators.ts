import { z } from 'zod'
import type { ICreatedAtFilterOperator } from '../field'

export const $eq = z.literal('$eq')
export const $neq = z.literal('$neq')
export const $contains = z.literal('$contains')
export const $not_contains = z.literal('$not_contains')
export const $starts_with = z.literal('$starts_with')
export const $ends_with = z.literal('$ends_with')
export const $regex = z.literal('$regex')

export const $is_true = z.literal('$is_true')
export const $is_false = z.literal('$is_false')

export const $in = z.literal('$in')
export const $nin = z.literal('$nin')

export const $gt = z.literal('$gt')
export const $lt = z.literal('$lt')
export const $gte = z.literal('$gte')
export const $lte = z.literal('$lte')

export const $start_eq = z.literal('$start_eq')
export const $start_neq = z.literal('$start_neq')
export const $start_gt = z.literal('$start_gt')
export const $start_lt = z.literal('$start_lt')
export const $start_gte = z.literal('$start_gte')
export const $start_lte = z.literal('$start_lte')

export const $end_eq = z.literal('$end_eq')
export const $end_neq = z.literal('$end_neq')
export const $end_gt = z.literal('$end_gt')
export const $end_lt = z.literal('$end_lt')
export const $end_gte = z.literal('$end_gte')
export const $end_lte = z.literal('$end_lte')

export const $is_empty = z.literal('$is_empty')
export const $is_not_empty = z.literal('$is_not_empty')

export const $is_today = z.literal('$is_today')
export const $is_not_today = z.literal('$is_not_today')
export const $is_tomorrow = z.literal('$is_tomorrow')
export const $is_yesterday = z.literal('$is_yesterday')
export const $between = z.literal('$between')

export const $has_file_type = z.literal('$has_file_type')
export const $has_file_extension = z.literal('$has_file_extension')

export const $is_root = z.literal('$is_root')

export const $is_me = z.literal('$is_me')
export const $is_not_me = z.literal('$is_not_me')

export const createdAtBuiltInOperators = new Set<ICreatedAtFilterOperator>([$is_today.value, $is_not_today.value])

export const operatorsWihtoutValue = z.union([
  $is_empty,
  $is_not_empty,
  $is_today,
  $is_not_today,
  $is_tomorrow,
  $is_yesterday,
  $is_root,
  $is_me,
  $is_not_me,
])
export const isOperatorWithoutValue = (value: string): boolean => operatorsWihtoutValue.safeParse(value).success
