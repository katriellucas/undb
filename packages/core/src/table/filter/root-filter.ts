import type { CompositeSpecification } from '@undb/domain'
import { ValueObject } from '@undb/domain'
import type { Option } from 'oxide.ts'
import { None } from 'oxide.ts'
import type { FieldId } from '../field'
import type { IGroup, IRootFilter } from './filter.js'
import { convertFilterSpec, isGroup } from './filter.js'

export class RootFilter extends ValueObject<IRootFilter> {
  get value() {
    return this.props
  }

  get group(): IGroup {
    if (Array.isArray(this.value)) {
      return { conjunction: '$and', children: this.value }
    }
    if (isGroup(this.value)) {
      return this.value
    }
    return { conjunction: '$and', children: [this.value] }
  }

  getSpec(userId: string): Option<CompositeSpecification> {
    return convertFilterSpec(this.value, userId)
  }

  // FIXME: remove field
  public removeField(fieldId: FieldId): Option<RootFilter> {
    return None
  }

  public toJSON() {
    return this.props
  }
}
