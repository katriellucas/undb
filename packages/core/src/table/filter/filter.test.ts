import { startOfYesterday } from 'date-fns'
import { convertFilterSpec, IRootFilter, rootFilter } from './filter.js'

const userId = 'usr1'

describe('test filters', () => {
  vi.setSystemTime(new Date(2020, 1, 1))

  test.each<IRootFilter>([
    {
      type: 'string',
      operator: '$eq',
      path: 'name',
      value: 'hello',
    },
    {
      type: 'string',
      operator: '$contains',
      path: 'name',
      value: 'hello',
    },
    {
      type: 'string',
      operator: '$starts_with',
      path: 'name',
      value: 'starts with',
    },
    {
      type: 'string',
      operator: '$ends_with',
      path: 'name',
      value: 'ends with',
    },
    {
      type: 'number',
      operator: '$neq',
      path: 'name',
      value: 1,
    },
    {
      type: 'number',
      operator: '$neq',
      path: 'name.nested',
      value: 1,
    },
    [
      {
        conjunction: '$and',
        children: [
          {
            type: 'string',
            operator: '$eq',
            path: 'name',
            value: 'hello',
          },
        ],
      },
    ],
    [
      {
        conjunction: '$or',
        children: [
          {
            type: 'string',
            operator: '$eq',
            path: 'name',
            value: 'hello',
          },
        ],
      },
    ],
    {
      type: 'number',
      operator: '$gt',
      path: 'gt',
      value: 1,
    },
    {
      type: 'number',
      operator: '$gte',
      path: 'gte',
      value: 1,
    },
    {
      type: 'number',
      operator: '$lt',
      path: 'lt',
      value: 1,
    },
    {
      type: 'number',
      operator: '$lte',
      path: 'lte',
      value: 1,
    },
    {
      type: 'date',
      operator: '$is_today',
      path: 'is_today',
      value: null,
    },
    {
      type: 'bool',
      operator: '$is_true',
      path: 'field',
      value: true,
    },
    {
      type: 'bool',
      operator: '$is_false',
      path: 'field',
      value: false,
    },
    {
      type: 'collaborator',
      operator: '$is_me',
      path: 'field',
      value: null,
    },
    {
      type: 'created-at',
      operator: '$is_tomorrow',
      path: 'is_tomorrow',
      value: null,
    },
    {
      type: 'updated-at',
      operator: '$is_tomorrow',
      path: 'is_tomorrow',
      value: null,
    },
    {
      type: 'date',
      operator: '$is_tomorrow',
      path: 'is_tomorrow',
      value: null,
    },
    {
      type: 'created-at',
      operator: '$is_yesterday',
      path: 'is_yesterday',
      value: null,
    },
    {
      type: 'updated-at',
      operator: '$is_yesterday',
      path: 'is_yesterday',
      value: null,
    },
    {
      type: 'date',
      operator: '$is_yesterday',
      path: 'is_yesterday',
      value: null,
    },
    {
      type: 'created-at',
      operator: '$between',
      path: 'between',
      value: [startOfYesterday().toISOString(), new Date().toISOString()],
    },
    {
      type: 'updated-at',
      operator: '$between',
      path: 'between',
      value: [startOfYesterday().toISOString(), new Date().toISOString()],
    },
    {
      type: 'date',
      operator: '$between',
      path: 'between',
      value: [startOfYesterday().toISOString(), new Date().toISOString()],
    },
  ])('should create root filter', (filter) => {
    const parsed = rootFilter.parse(filter)
    expect(parsed).toEqual(filter)

    const spec = convertFilterSpec(filter, userId)
    expect(spec).toMatchSnapshot()
  })
})
