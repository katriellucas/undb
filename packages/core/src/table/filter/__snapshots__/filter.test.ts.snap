// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`test filters > should create root filter 1`] = `
OptionType {
  Symbol(T): true,
  Symbol(Val): StringEqual {
    "fieldId": "name",
    "value": StringFieldValue {
      "props": {
        "value": "hello",
      },
    },
  },
}
`;

exports[`test filters > should create root filter 2`] = `
OptionType {
  Symbol(T): true,
  Symbol(Val): StringContain {
    "fieldId": "name",
    "value": StringFieldValue {
      "props": {
        "value": "hello",
      },
    },
  },
}
`;

exports[`test filters > should create root filter 3`] = `
OptionType {
  Symbol(T): true,
  Symbol(Val): StringStartsWith {
    "fieldId": "name",
    "value": StringFieldValue {
      "props": {
        "value": "starts with",
      },
    },
  },
}
`;

exports[`test filters > should create root filter 4`] = `
OptionType {
  Symbol(T): true,
  Symbol(Val): StringEndsWith {
    "fieldId": "name",
    "value": StringFieldValue {
      "props": {
        "value": "ends with",
      },
    },
  },
}
`;

exports[`test filters > should create root filter 5`] = `
OptionType {
  Symbol(T): true,
  Symbol(Val): m {
    "spec": NumberEqual {
      "fieldId": "name",
      "value": NumberFieldValue {
        "props": {
          "value": 1,
        },
      },
    },
  },
}
`;

exports[`test filters > should create root filter 6`] = `
OptionType {
  Symbol(T): true,
  Symbol(Val): m {
    "spec": NumberEqual {
      "fieldId": "name.nested",
      "value": NumberFieldValue {
        "props": {
          "value": 1,
        },
      },
    },
  },
}
`;

exports[`test filters > should create root filter 7`] = `
OptionType {
  Symbol(T): true,
  Symbol(Val): StringEqual {
    "fieldId": "name",
    "value": StringFieldValue {
      "props": {
        "value": "hello",
      },
    },
  },
}
`;

exports[`test filters > should create root filter 8`] = `
OptionType {
  Symbol(T): true,
  Symbol(Val): StringEqual {
    "fieldId": "name",
    "value": StringFieldValue {
      "props": {
        "value": "hello",
      },
    },
  },
}
`;

exports[`test filters > should create root filter 9`] = `
OptionType {
  Symbol(T): true,
  Symbol(Val): NumberGreaterThan {
    "fieldId": "gt",
    "value": NumberFieldValue {
      "props": {
        "value": 1,
      },
    },
  },
}
`;

exports[`test filters > should create root filter 10`] = `
OptionType {
  Symbol(T): true,
  Symbol(Val): NumberGreaterThanOrEqual {
    "fieldId": "gte",
    "value": NumberFieldValue {
      "props": {
        "value": 1,
      },
    },
  },
}
`;

exports[`test filters > should create root filter 11`] = `
OptionType {
  Symbol(T): true,
  Symbol(Val): NumberLessThan {
    "fieldId": "lt",
    "value": NumberFieldValue {
      "props": {
        "value": 1,
      },
    },
  },
}
`;

exports[`test filters > should create root filter 12`] = `
OptionType {
  Symbol(T): true,
  Symbol(Val): NumberLessThanOrEqual {
    "fieldId": "lte",
    "value": NumberFieldValue {
      "props": {
        "value": 1,
      },
    },
  },
}
`;

exports[`test filters > should create root filter 13`] = `
OptionType {
  Symbol(T): true,
  Symbol(Val): DateIsToday {
    "fieldId": "is_today",
    "value": DateFieldValue {
      "props": {
        "value": null,
      },
    },
  },
}
`;

exports[`test filters > should create root filter 14`] = `
OptionType {
  Symbol(T): true,
  Symbol(Val): BoolIsTrue {
    "fieldId": "field",
    "value": BoolFieldValue {
      "props": {
        "value": true,
      },
    },
  },
}
`;

exports[`test filters > should create root filter 15`] = `
OptionType {
  Symbol(T): true,
  Symbol(Val): BoolIsFalse {
    "fieldId": "field",
    "value": BoolFieldValue {
      "props": {
        "value": false,
      },
    },
  },
}
`;

exports[`test filters > should create root filter 16`] = `
OptionType {
  Symbol(T): true,
  Symbol(Val): CollaboratorEqual {
    "fieldId": "field",
    "value": CollaboratorFieldValue {
      "props": [
        "usr1",
      ],
    },
  },
}
`;

exports[`test filters > should create root filter 17`] = `
OptionType {
  Symbol(T): true,
  Symbol(Val): DateIsTomorrow {
    "fieldId": "is_tomorrow",
    "value": DateFieldValue {
      "props": {
        "value": null,
      },
    },
  },
}
`;

exports[`test filters > should create root filter 18`] = `
OptionType {
  Symbol(T): true,
  Symbol(Val): DateIsTomorrow {
    "fieldId": "is_tomorrow",
    "value": DateFieldValue {
      "props": {
        "value": null,
      },
    },
  },
}
`;

exports[`test filters > should create root filter 19`] = `
OptionType {
  Symbol(T): true,
  Symbol(Val): DateIsTomorrow {
    "fieldId": "is_tomorrow",
    "value": DateFieldValue {
      "props": {
        "value": null,
      },
    },
  },
}
`;

exports[`test filters > should create root filter 20`] = `
OptionType {
  Symbol(T): true,
  Symbol(Val): DateIsYesterday {
    "fieldId": "is_yesterday",
    "value": DateFieldValue {
      "props": {
        "value": null,
      },
    },
  },
}
`;

exports[`test filters > should create root filter 21`] = `
OptionType {
  Symbol(T): true,
  Symbol(Val): DateIsYesterday {
    "fieldId": "is_yesterday",
    "value": DateFieldValue {
      "props": {
        "value": null,
      },
    },
  },
}
`;

exports[`test filters > should create root filter 22`] = `
OptionType {
  Symbol(T): true,
  Symbol(Val): DateIsYesterday {
    "fieldId": "is_yesterday",
    "value": DateFieldValue {
      "props": {
        "value": null,
      },
    },
  },
}
`;

exports[`test filters > should create root filter 23`] = `
OptionType {
  Symbol(T): true,
  Symbol(Val): DateBetween {
    "date1": 2020-01-31T00:00:00.000Z,
    "date2": 2020-02-01T00:00:00.000Z,
    "fieldId": "between",
  },
}
`;

exports[`test filters > should create root filter 24`] = `
OptionType {
  Symbol(T): true,
  Symbol(Val): DateBetween {
    "date1": 2020-01-31T00:00:00.000Z,
    "date2": 2020-02-01T00:00:00.000Z,
    "fieldId": "between",
  },
}
`;

exports[`test filters > should create root filter 25`] = `
OptionType {
  Symbol(T): true,
  Symbol(Val): DateBetween {
    "date1": 2020-01-31T00:00:00.000Z,
    "date2": 2020-02-01T00:00:00.000Z,
    "fieldId": "between",
  },
}
`;
